CVLByExamples/ConstantProductPoolExample/certora/spec/ConstantProductPool.spec

```diff
@@ -0,0 +1,264 @@
+*/
+
+rule integrityOfSwap(address recipient) {
+    env e; /* represents global solidity variables such as msg.sender, block.timestamp */
+    setup(e);
+    require recipient != currentContract; /* currentContract is a CVL keyword, assigned the main contract under test */  
+    uint256 balanceBefore = _token0.balanceOf(recipient);
+    uint256 amountOut = swap(_token1, recipient);
+    uint256 balanceAfter = _token0.balanceOf(recipient);
+    assert to_mathint(balanceAfter) == balanceBefore + amountOut; 
+}
+
+/*
+Property: Only the user itself or an allowed spender can decrease the user's LP balance.
+
+This property is implemented as a parametric rule - it checks all public/external methods of the contract.
+
+This property catches a bug in which there is a switch between the token and the recipient in burnSingle:
+        transfer( recipient, tokenOut, amountOut);
+
+Formula:
+        { b = balanceOf(account), allowance = allowance(account, e.msg.sender) }
+            op by e.msg.sender;
+        { balanceOf(account) < b =>  (e.msg.sender == account  ||  allowance >= (before-balanceOf(account)) }
+*/
+
+rule noDecreaseByOther(method f, address account) {
+    env e;
+    setup(e);
+    require e.msg.sender != account;
+    require account != currentContract; 
+    uint256 allowance = allowance(account, e.msg.sender); 
+    
+    uint256 before = balanceOf(account);
+    calldataarg args;
+    f(e,args); /* check on all possible arguments */
+    uint256 after = balanceOf(account);
+    /* logic implication : true when: (a) the left hand side is false or (b) right hand side is true  */
+    assert after < before =>  (e.msg.sender == account  ||  to_mathint(allowance) >= (before-after))  ;
+}
+
+
+/*
+Property: For both token0 and token1 the balance of the system is at least as much as the reserves.
+
+This property is implemented as an invariant. 
+Invariants are a specification of a condition that should always be true once an operation is concluded.
+In addition, the invariant also checks that it holds right after the constructor of the code runs.
+
+This invariant also catches the bug in which there is a switch between the token and the recipient in burnSingle:
+        transfer( recipient, tokenOut, amountOut);
+
+Formula:
+    getReserve0() <= _token0.balanceOf(currentContract) &&
+    getReserve1() <= _token1.balanceOf(currentContract)
+*/
+
+invariant balanceGreaterThanReserve()
+    (getReserve0() <= _token0.balanceOf(currentContract))&&
+    (getReserve1() <= _token1.balanceOf(currentContract))
+    {
+        preserved with (env e){
+         setup(e);
+        }
+    }
+
+
+/*
+Property: Integrity of totalSupply with respect to the amount of reserves. 
+
+This is a high level property of the system - the ability to pay back liquidity providers.
+If there are any LP tokens (the totalSupply is greater than 0), then neither reserves0 nor reserves1 should ever become zero (otherwise the pool could not produce the underlying tokens).
```

Comment:
```suggestion
If there are any LP tokens (i.e. if the total supply is greater than 0), then neither `reserves0` nor `reserves1` should ever become zero (otherwise the pool could not produce the underlying tokens).
```